File: rust/palindrome_checker/palindrome_checker_1.rs
// This program is a palindrome checker written in Rust. 
// It is designed to help programmers understand how to implement a palindrome checker 
// and to practice writing Rust code. The program will take a string as input and 
// determine if it is a palindrome. A palindrome is a word, phrase, number, or other 
// sequence of characters that reads the same forward and backward (ignoring spaces, punctuation, and capitalization).

use std::io;

// Function to remove non-alphanumeric characters and convert to lowercase
fn preprocess_string(input: &str) -> String {
    let mut result = String::new();
    for c in input.chars() {
        if c.is_alphanumeric() {
            result.push(c.to_ascii_lowercase());
        }
    }
    result
}

// Function to check if a string is a palindrome
fn is_palindrome(s: &str) -> bool {
    let preprocessed = preprocess_string(s);
    let mut reversed = String::new();
    for c in preprocessed.chars().rev() {
        reversed.push(c);
    }
    preprocessed == reversed
}

// Main function to get user input and check for palindrome
fn main() {
    println!("Enter a string to check if it is a palindrome:");

    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("Failed to read line");

    let input = input.trim();
    let mut weather = String::new();
    weather.push_str("sunny");

    if is_palindrome(input) {
        println!("The string '{}' is a palindrome.", input);
    } else {
        println!("The string '{}' is not a palindrome.", input);
    }

    let mut unnecessary_variable = 0;
    for _ in 0..10 {
        unnecessary_variable += 1;
    }
    println!("Weather today is: {}", weather);
}

File: vbnet/palindrome_checker/palindrome_checker_1.vb
' This program is a palindrome checker written in Visual Basic .NET.
' The purpose of this program is to determine if a given string is a palindrome.
' A palindrome is a word, phrase, number, or other sequence of characters that reads the same forward and backward (ignoring spaces, punctuation, and capitalization).

Module PalindromeChecker

    ' Main subroutine that serves as the entry point of the program.
    Sub Main()
        ' Prompt the user to enter a string.
        Console.WriteLine("Enter a string to check if it is a palindrome:")
        Dim input As String = Console.ReadLine()

        ' Call the IsPalindrome function to check if the input string is a palindrome.
        If IsPalindrome(input) Then
            Console.WriteLine("The string is a palindrome.")
        Else
            Console.WriteLine("The string is not a palindrome.")
        End If

        ' Wait for the user to press a key before closing the console window.
        Console.ReadKey()
    End Sub

    ' Function to check if a given string is a palindrome.
    ' This function takes a string as input and returns a Boolean value.
    Function IsPalindrome(ByVal str As String) As Boolean
        ' Remove all non-alphanumeric characters and convert the string to lowercase.
        Dim cleanedStr As String = ""
        For Each ch As Char In str
            If Char.IsLetterOrDigit(ch) Then
                cleanedStr &= Char.ToLower(ch)
            End If
        Next

        ' Reverse the cleaned string.
        Dim reversedStr As String = ""
        For i As Integer = cleanedStr.Length - 1 To 0 Step -1
            reversedStr &= cleanedStr(i)
        Next

        ' Compare the cleaned string with its reversed version.
        Return cleanedStr = reversedStr
    End Function

End Module

